### DISCUSSION TITLE
Rails x RSpec x GraphQLでの単体テストの最適解を考えたいスレッド

### DISCUSSION THREAD

**Discussion by DaisukeShinoku** [yuki-snow1823/diary#50] [Ideas]
頑張ろう

---

7/14(金) 追記

最初に新奥の方で[PR](https://github.com/yuki-snow1823/diary/pull/58)を作成したんだけど、チームとして決めることがたくさんあるのでDiscussionで進めているよ

7/20(木) 20:00 に [meet](https://meet.google.com/mza-zjnq-sdk) で同期的に決めていくのでみんな入ってきてくれい

**Comment by DaisukeShinoku** [comment]
以下のQiita記事は参考になりそう
[GraphQLのテストいろいろ](https://qiita.com/Paul_Dirac/items/a99c40c5f74efe76d9ec)

とりあえず一番楽そうな「resolverだけ叩く方法」でいいかな、と思っているけど、そこから改めて考えてみるのが良いのかも？

**Reply by yuki-snow1823** [reply]
3つ分かれてるの面白いですね。そしてどれも適切なものをテストしている気がします。
叩いた時にどうか > 実行した時にどうか > resolverでどうかの順に規模が大きいですね。
叩いた時にどうかはテストしてる範囲も必然的に広そうです。

叩いた時にどうか、はそもそも動作するかとか実際にテストしたい以外の部分も見てしまうと思うので、方針としてresolverのみというのは賛成です。ちょっとexecuteの方も良いとは思うので調べてみます。

**Comment by yuki-snow1823** [comment]
そう考えると、modelやrequest specではそもそもないので
spec/graphql/mutationsとかのディレクトリ作ることを提案します🙋‍♂️

**Reply by DaisukeShinoku** [reply]
ではそれで！
あとはテストの個別の書き方をどうするのが良いかか・・・（ナニモワカラナイ）

**Reply by yuki-snow1823** [reply]
@DaisukeShinoku 
CreateJournalの場合ですが
- resolveメソッドを実行する場合
  - userのidがあればjournalオブジェクト作成されて、resolverの戻り値がuserのidパラメータを持っていること
  - userのidがなければ作成されないこと 
  - titleがあれば同上
  - titleがなければ作成されないこと
  - contentがあれば同上
  - contentがなければ作成されないこと

みたいな感じかなぁと思いました。

動作確認してないですが、こんな感じですかね。

```ruby
RSpec.describe Mutations::CreateJournal, type: :request do
  contexct 'userが存在する場合' do
    it 'userのidを持つjournalが作られること' do
      user = create(:user)
      mutation = Mutations::CreateJournal.new(field: nil, object: nil, context: { current_user: user })
      title = 'Test Title'
      content = 'test journal.'

      ret = mutation.resolve(title: title, content: content, user_id: user.id)
      journal = Journal.first

      expect(journal).to have_attributes(title: title, content: content, user_id: user.id)
      expect(journal.user_id).to eq user.id
    end
  end
end
```

ただこれだとコード量が増えるので、FactoryBotで、titleとcontextとuserを持っているJournalを作っておいて
各テストケースで必要に応じてnilを与えたりすればいいと思いました。
このテストだと、titleとcontextを勝手に補っちゃっているので、それはFactoryの方でやればいいかなと考えました。

**Reply by DaisukeShinoku** [reply]
@yuki-snow1823 

> ただこれだとコード量が増えるので、FactoryBotで、titleとcontextとuserを持っているJournalを作っておいて
各テストケースで必要に応じてnilを与えたりすればいいと思いました。

ここの意味するところがちょっとわかってないかもです！

**Reply by DaisukeShinoku** [reply]
> FactoryBotで、titleとcontextとuserを持っているJournalを作っておいて

文字通りに捉えるとresolverを使わないように読めてしまっている（modelのテストみたいに思える）

**Reply by yuki-snow1823** [reply]
@DaisukeShinoku 
寝ぼけてましたw

FactoryBotじゃないですね。
letでした！

こんな感じで、基本titleもcontentもuserも含まれるletを作成しておいて
contextごとでテストしたいケースだけnilで変数を置き換えるイメージでした。

```graphql
RSpec.describe Mutations::CreateJournal, type: :request do
  describe 'CreateJournal mutation' do
    let(:mutation) { Mutations::CreateJournal.new(field: nil, object: nil, context: { current_user: user }) }
    let(:title) { 'Test Title' }
    let(:content) { 'test journal.' }
    let(:resolve_params) { { title: title, content: content, user_id: create(:user) } }
    let(:resolve) { mutation.resolve(resolve_params) }
    let(:journal) { Journal.first }

    context 'userが存在する場合' do
      let(:user) { create(:user) }

      it 'userのidを持つjournalが作られること' do
        resolve
        expect(journal).to have_attributes(title: title, content: content, user_id: user.id)
        expect(journal.user_id).to eq user.id
      end
    end

    context 'userが存在しない場合' do
      let(:user) { nil }

      it 'journalが作成されないこと' do
        expect { resolve }.to raise_error(StandardError) # ここはgraphql-rubyのエラーの実装によります
      end
    end
  end
end

```

**Comment by DaisukeShinoku** [comment]
@yuki-snow1823 

意図理解です！
DRYの仕方とか、contextの分け方は当日のモブで会話するのが良さそう
最初 it で全部書いてそこから共通化できるところを切り出すイメージ

---

`ここはgraphql-rubyのエラーの実装によります`

どっちかというとこっちがアイディアが多く欲しい部分ですかね〜
いったん自分の方でも仮実装を用意しようかと思いますが、あんまり良い方法は思い付いていないです・・・

**Reply by yuki-snow1823** [reply]
@DaisukeShinoku 
いいですね！個人的にはテキストファイルにテストケース書き出してみるとか、ディシジョンテーブルみたいなところにまるばつ作ってケースを網羅してから書き出せると良いかなと思いました🙆‍♂️

ディシジョンテーブル
https://shiftasia.com/ja/column/%E3%83%87%E3%82%B7%E3%82%B8%E3%83%A7%E3%83%B3%E3%83%86%E3%83%BC%E3%83%96%E3%83%AB%EF%BC%88%E6%B1%BA%E5%AE%9A%E8%A1%A8%EF%BC%89%E3%81%A8%E3%81%AF/

エラーについて考えてみますね。
executeのほうだと、GraphQL::ExecuteErrorみたいなやつがあったと思うんですが、今回のresolveの場合は初めてなので考えてみます！

**Reply by yuki-snow1823** [reply]
多分なのですが、ユーザーIDとかを渡さないと普通にActiveRecordのエラーが起こりそうですよね。

バリデーションのテストはモデルスペックに書けばいいので、GraphQLの方ではざっくりとエラーが起きたかどうかのテストみたいなのを書けばいいかなーって思いました。

**Reply by yuki-snow1823** [reply]
引数のuserを無くしたところ、NoMethodErrorが出ました！

```
Failures:

  1) Mutations::CreateJournal CreateJournal mutation userが存在しない場合 journalが作成されないこと
     Failure/Error: let(:resolve) { mutation.resolve(title: title, content: content, user_id: user.id) }
     
     NoMethodError:
       undefined method `id' for nil:NilClass
     # ./spec/graphql/create_journal_spec.rb:9:in `block (3 levels) in <top (required)>'
     # ./spec/graphql/create_journal_spec.rb:24:in `block (4 levels) in <top (required)>'
```

```ruby
require 'rails_helper'

RSpec.describe Mutations::CreateJournal, type: :request do
  describe 'CreateJournal mutation' do
    let(:mutation) { Mutations::CreateJournal.new(field: nil, object: nil, context: { current_user: user }) }
    let(:title) { 'Test Title' }
    let(:content) { 'test journal.' }
    let(:user) { create(:user) }
    let(:resolve) { mutation.resolve(title: title, content: content, user_id: user.id) }
    let(:journal) { Journal.first }

    context 'userが存在する場合' do
      it 'userのidを持つjournalが作られること' do
        resolve
        expect(journal).to have_attributes(title: title, content: content, user_id: user.id)
        expect(journal.user_id).to eq user.id
      end
    end

    context 'userが存在しない場合' do
      let(:user) { nil }

      it 'journalが作成されないこと' do
        p resolve
        expect { resolve }.to raise_error(StandardError)
      end
    end
  end
end
```

StandardErrorが継承されているので、こうすればテストは通りますね。1案です。
なんのエラーかはわからないが、作成されないし不適切な引数だとエラーが起こることはテストできますね。

**Reply by DaisukeShinoku** [reply]
[別コメント](https://github.com/yuki-snow1823/diary/discussions/50#discussioncomment-6450110)に載っけたのですが、executeとresoleverで仮実装しました

> GraphQLの方ではざっくりとエラーが起きたかどうかのテストみたいなのを書けばいいかなーって思いました。

上記前提ならresoleverでも問題ないのですが、executeの方が自由度は高そうと思いいい意味で悩んできた、後からやりたいこと増えた時にresolverだと限界があるのかもしれない？

※ 新奥がErrorを検知以外のテストの仕方をわかっていないだけで、他にもいろいろできるのかもしれませんが

**Reply by yuki-snow1823** [reply]
>resolverだと限界があるのかもしれない

エラーのテストに限界ありそうですよね。
executeのRSpecは普段書いてるのですが、GraphQLのエラーを拡張してエラーのステータスを実装してそれをテストしたり
色々できることあります！

仮実装感謝です！これも議論して分かったことですね〜

前これ実務でも使って実装＋テスト書きました
GraphQLにおけるエラーハンドリングの仕方 - ZOZO TECH BLOG https://techblog.zozo.com/entry/graphql_error_handling

**Comment by DaisukeShinoku** [comment]
モブプロ用にDRYとかはいったん無視したPRを２つ作成

[mutations_rspec仮実装(execute方式)](https://github.com/yuki-snow1823/diary/pull/73)

[mutations_rspec仮実装(resolver方式) ](https://github.com/yuki-snow1823/diary/pull/74)

**Reply by yuki-snow1823** [reply]
![image](https://github.com/yuki-snow1823/diary/assets/59280290/d9a55089-6887-45a9-a82d-ac673bcd0406)

**Comment by yuki-snow1823** [comment]
@DaisukeShinoku さんとお話ししました
- より明確にエラーのテストをするためにExecuteの方でテストを書くことにしました
- 例えば、他の人のものを更新できない（認可）の実装をしたときに、認可のテストを書いていただく
- ログインしていなければ投稿できない（認証）も同様
- 今はそれらのテストを追加せず、以下をテストする
  - 正常系
  - 異常系はモデルスペックと被るし、上記の理由でまだ機能の方ができていないので正常系だけでいい気がする

**Comment by yuki-snow1823** [comment]
NEXT: 
- updateの方に何を書くか
  - updateのエラーのテストケースを入れておくことで、エラーのテストが今後書きやすくなる（読めばわかる状態）
  - `result.dig('errors').first.dig('message').to be_present`みたいになりそうですね
  - 例えば、updateのjournal_idを適当に送るとか
    - Journal.last + 1 のIDを入れるとかどうでしょう 

- deleteの方に何を書くか
  - before do で消したいレコードを作る気がしますね
  - `.to change(Journal, :count).by(-1)`よさみざわです →これだけでいいと思います

**Reply by yuki-snow1823** [reply]
@DaisukeShinoku お願いします🙏

**Reply by DaisukeShinoku** [reply]
@yuki-snow1823 
PR ready for reviewです！
https://github.com/yuki-snow1823/diary/pull/73

**Reply by yuki-snow1823** [reply]
@DaisukeShinoku 
レビューしました！
